tests/contrib/gp/test_models.py

```diff
@@ -71,29 +98,195 @@ def test_model_forward(model_class, X, y, kernel, likelihood):
+
+
+@pytest.mark.init(rng_seed=0)
+@pytest.mark.disable_validation()
+def test_inference_svgp():
+    N = 1000
+    X = dist.Uniform(torch.zeros(N), torch.ones(N)*5).sample()
+    y = 0.5 * torch.sin(3*X) + dist.Normal(torch.zeros(N), torch.ones(N)*0.5).sample()
+    kernel = RBF(input_dim=1)
+    Xu = torch.linspace(0, 5, 10)
+
+    svgp = SparseVariationalGP(X, y, kernel, Xu, Gaussian())
+    svgp.optimize(optim.Adam({"lr": 0.01}), num_steps=1000)
+
+    Xnew = torch.linspace(0, 5, 100)
+    loc, var = svgp(Xnew, full_cov=False)
+    target = 0.5 * torch.sin(3*Xnew)
+
+    assert_equal((loc - target).abs().mean().item(), 0, prec=0.05)
+
+
+@pytest.mark.parametrize("model_class, X, y, kernel, likelihood", TEST_CASES, ids=TEST_IDS)
+@pytest.mark.disable_validation()
+def test_inference_with_empty_latent_shape(model_class, X, y, kernel, likelihood):
+    # regression models don't use latent_shape (default=torch.Size([]))
+    if model_class is GPRegression or model_class is SparseGPRegression:
+        return
+    elif model_class is VariationalGP:
+        gp = model_class(X, y, kernel, likelihood, latent_shape=torch.Size([]))
+    else:  # model_class is SparseVariationalGP
+        gp = model_class(X, y, kernel, X, likelihood, latent_shape=torch.Size([]))
+
+    gp.optimize(num_steps=1)
+
+
+@pytest.mark.parametrize("model_class, X, y, kernel, likelihood", TEST_CASES, ids=TEST_IDS)
+@pytest.mark.disable_validation()
+def test_hmc(model_class, X, y, kernel, likelihood):
     if model_class is SparseGPRegression or model_class is SparseVariationalGP:
         gp = model_class(X, y, kernel, X, likelihood)
     else:
         gp = model_class(X, y, kernel, likelihood)
 
-    gp.optimize(num_steps=1)
+    kernel.set_prior("variance", dist.Uniform(torch.tensor(0.5), torch.tensor(1.5)))
+    kernel.set_prior("lengthscale", dist.Uniform(torch.tensor(1), torch.tensor(3)))
+
+    hmc_kernel = HMC(gp.model, step_size=1)
+    mcmc_run = MCMC(hmc_kernel, num_samples=10)
+
+    post_trace = defaultdict(list)
+    for trace, _ in mcmc_run._traces():
+        variance_name = pyro.param_with_module_name(kernel.name, "variance")
+        post_trace["variance"].append(trace.nodes[variance_name]["value"])
+        lengthscale_name = pyro.param_with_module_name(kernel.name, "lengthscale")
+        post_trace["lengthscale"].append(trace.nodes[lengthscale_name]["value"])
+        if model_class is VariationalGP:
+            f_name = pyro.param_with_module_name(gp.name, "f")
+            post_trace["f"].append(trace.nodes[f_name]["value"])
+        if model_class is SparseVariationalGP:
+            u_name = pyro.param_with_module_name(gp.name, "u")
+            post_trace["u"].append(trace.nodes[u_name]["value"])
+
+    for param in post_trace:
+        param_mean = torch.mean(torch.stack(post_trace[param]), 0)
+        logger.info("Posterior mean - {}".format(param))
+        logger.info(param_mean)
+
+
+@pytest.mark.disable_validation()
```

Comment:
Why does this disable validation?

Reply:
@fritzo Enable it will give me [error](https://gist.github.com/fehiepsi/00f7b2fdfcc7bf39ee25968bad40f6db) which I mentioned in https://github.com/uber/pyro/pull/936 . It seems that I have to use `pyro.iarange` for a bunch of independent dimensions? How can I handle its ndim automatically?
```
pyro.sample(dist.Normal(torch.zeros(3), torch.ones(3)), obs=torch.rand(9, 8, 7, 6, 3))
```
Another way is to use extra_event_dim, but you put many questions why so at #888. So I thought what I did was wrong. :(

Reply:
@fehiepsi You may have done it right in #888. We're still working out best practices. Let's enable validation and figure out whether you need to use `iarange` or `extra_event_dims`.

Reply:
> How can I handle its ndim automatically?

Good question. A manual answer to your example is:
```py
with iarange('a'), iarange('b'), iarange('c'), iarange('d'):
    pyro.sample(dist.Normal(torch.zeros(3), torch.ones(3)), obs=torch.rand(9, 8, 7, 6, 3))
```
I guess one answer as to how to do this automatically is to put the responsibility with the code that introduced each such independent dimension. Would this work for you, e.g. by putting `iarange`s in your tests that call GP code, and expecting that users of pyro.contrib.gp would use `iarange` when needed, outside of pyro.contrib.gp?

Reply:
Yes, let's temporarily use "extra_event_dims". I will mark TODOs at points needed to be revised (if we follow `iarange` approach).